google-cloud-bigtable/lib/google/cloud/bigtable/table/column_families_updater.rb

```diff
@@ -0,0 +1,168 @@
+# frozen_string_literal: true
+
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+module Google
+  module Cloud
+    module Bigtable
+      class Table
+        ##
+        # An accumulator for modifications to a table's column families. See
+        # {Table#column_families}. Modifications will be atomically applied to
+        # the table's column families. Entries are applied in order, meaning
+        # that earlier modifications can be masked by later ones (in the case
+        # of repeated updates to the same family, for example).
+        #
+        # @example
+        #   require "google/cloud/bigtable"
+        #
+        #   bigtable = Google::Cloud::Bigtable.new
+        #
+        #   table = bigtable.table("my-instance", "my-table", perform_lookup: true)
+        #
+        #   table.column_families do |cfs|
+        #
+        #     cfs.add "cf1", Google::Cloud::Bigtable::GcRule.max_age(600)
+        #     cfs.add "cf2", Google::Cloud::Bigtable::GcRule.max_versions(5)
+        #
+        #     rule_1 = Google::Cloud::Bigtable::GcRule.max_versions(3)
+        #     rule_2 = Google::Cloud::Bigtable::GcRule.max_age(600)
+        #     rule_union = Google::Cloud::Bigtable::GcRule.union(rule_1, rule_2)
+        #     cfs.update "cf3", rule_union
+        #
+        #     cfs.delete "cf5"
+        #   end
+        #
+        class ColumnFamiliesUpdater
```

Comment:
Thoughts on naming this `class ColumnFamiliesMap::Updater`, matching its use in other libraries?

Reply:
And inheriting from `ColumnFamiliesMap?

Reply:
That's actually what I started with, and I can move it back. But the nested class in the Hash delegate just seemed a little confusing.

Reply:
I thought `ColumnFamiliesMap` was no longer going to be a Hash delegate.. Is that no longer happening? Or is that in a future change?

Reply:
After splitting the `ColumnFamilyMap` and the updater, so much complexity was removed from our plan for a new `ColumnFamilyMap` that I no longer felt the need to change it much.

Reply:
If we've gone that far, then here's another question. The only real reason now for `ColumnFamilyMap` to be an actual class rather than simply a frozen `Hash`, is the `add` method. Why not break out a `ColumnFamiliesCreator` class instead?

Reply:
I kept `ColumnFamilyMap` as a `Hash` delegate simply to preserve the existing API as much as possible where not problematic. I have no problem changing the object yielded by `create_table` to a simple type with just an `add` method. I also agree that `Table.column_families` should just return an ordinary frozen `Hash`.

Reply:
Replaced `ColumnFamilyMap` with `ColumnFamiliesCreator`.

Reply:
@dazuma Now we have 3 classes that represent column families, ColumnFamilyMap (Hash delegate), ColumnFamilyUpdater, and ColumnFamilyCreator. I think this is confusing.

I propose that we have just one ColumnFamilyMap class that does not allow modifications when frozen. The object can be created by calling `ColumnFamilyMap.new` without any arguments, and when calling `ColumnFamilyMap#modifications` it returns the Mutation rules to create the entire map. Also, when created by calling the private `ColumnFamilyMap.from_grpc` it will track the changes made and return them when calling `ColumnFamilyMap#modifications`. This would allow users to create new objects to pass when creating new tables, and also allow for `Table#column_families` to yield an object that can track be used to change them. It would be the same object either way.

Thoughts?

Reply:
Would some mechanism prevent calling `update` and `delete` when the class is used for creating new tables?

Reply:
Also, I think mixing the `ColumnFamiliesUpdater` methods in with a hash might be problematic. The hash methods won't correctly reflect changes made, since those changes are just accumulated modifications in the updater. They aren't yet changes to `Table#column_families`.

Reply:
I don't understand why we would want to prevent users from calling `#update` and `#delete`.

Reply:
We don't have three classes. Just one: only `ColumnFamilyMap` represents a set of column families. The other two classes are better thought of as part of the service interface. (Notice the only time they appear is as an argument to a block.)

I'm not married to this design. I just think we should do something to avoid the confusion that happens when an object provides methods that don't make sense or have ambiguous semantics. For example, when is it ever okay to use the Hash mutation methods like `[]=`? If we allow it during create, we open the door to a lot of usage errors (e.g. key and name not matching). If we allow it during update, we're faced with the daunting task of inferring a series of legal column table modifications from arbitrary hash operations.

Or, as Chris mentions, we have three use cases (initially build data for table creation, update existing table, and get current state), and methods like `update` and `delete` and `add` that each apply to only one of the three use cases, so now we have to document carefully and probably put in runtime checks to ensure the user doesn't try to do something we can't allow in a particular context.

Do we allow the user to write things like the following? What are the semantics? If they're not allowed, how do we communicate that to users? If you think it's better to recombine it all into a single class, I'm not opposed to that but I do think we'd need to document very carefully what can be called under what circumstance.

```ruby
cfs = my_table.column_families
cfs["family2"] = ColumnFamily.new(...)
cfs.delete("family3")
```

```ruby
cfs = ColumnFamilyMap.new
cfs["family1"] = ColumnFamily.new(...)
cfs.delete("family1")
Table.new(column_families: cfs)
```

```ruby
Table.new do |cfs|
  cfs["family1"] = cfs["family2"]
end
```

Reply:
Here's a hypothetical example to show why I don't think it's very easy to mix hash methods with the updater, which just accumulates modifications. It would be very difficult to reflect these modifications in the separate collection of `ColumnFamily` objects.

```rb
table.column_families do |cf_updater|   
  cf_updater.size #=> 3  
  cf_updater.add "cf4", Google::Cloud::Bigtable::GcRule.max_versions(5)  
  cf_updater.size #=> 3  # never changes  
  cf_updater.delete "cf3" 
  cf_updater["cf3"] #=> <ColumnFamily name: "cf3" ...>
 end
```

Reply:
> We don't have three classes. Just one: only `ColumnFamilyMap`

Currently, this branch does not have a `ColumnFamilyMap` class, as it has been removed. `Table#column_families` returns a frozen `Hash`, `Table#column_families` yields a `ColumnFamilyUpdater` object, `Project#create_table` accepts a Hash (how does a user create a Hash with ColumnFamily objects anyway?), and `Project#create_table` yields a `ColumnFamiliesCreator` object